File: Week-1-C/credit/credit.c
#include <stdio.h>

long get_card_number(void);
int get_length_of_card_number(long card_number);
int calculate_checksum(long card_number);
int get_starting_numbers(long card_number);
void validate_card_number(long card_number, int length);

int main(void) {
    
    long int card_number = get_card_number();
    int length = get_length_of_card_number(card_number);

    validate_card_number(card_number, length);

    return 0;
}

// Returns valid card number input from user
long get_card_number(void)
{
    long int card_number;
    int result;
    do
    {
      printf("Number: ");
      result = scanf("%li", &card_number);
      if (result != 1)
      {
          while (getchar() != '\n');
      }
    }
    while (result != 1);
    return card_number;
}

// Returns the length of the card number
int get_length_of_card_number(long card_number)
{
    int length = 0;
    if (card_number == 0) length = 1;
    else
    {
        while (card_number != 0)
        {
            card_number /= 10;
            length++;
        }
    }
    return length;
}

// Returns a sum from checksum algorithm to determine is last digit of sum is 0
int calculate_checksum(long card_number)
{
    int normal_sum = 0;
    int doubled_sum = 0;
    int should_double = 0;

    while (card_number != 0)
    {
        int num = card_number % 10;

        if (should_double)
        {
            int num_doubled = num * 2;
            if (num_doubled > 9)
            {
                doubled_sum += num_doubled % 10;
                doubled_sum += num_doubled / 10;
            }
            else
            {
                doubled_sum += num_doubled;
            }
        }
        else
        {
            normal_sum += num;
        }

        card_number /= 10;
        should_double = !should_double;
    }
    return normal_sum + doubled_sum;
}

// Return the starting numbers to compare to valid card starting numbers
int get_starting_numbers(long card_number)
{
    int length = get_length_of_card_number(card_number);

    if (length == 16 || length == 15)
    {
        for (int i = 0; i < length-2; i++) card_number /= 10;
        return card_number;
    }
    else if (length == 13)
    {
        for (int i = 0; i < length-1; i++) card_number /= 10;
        return card_number;
    }
    else return 0;
}

// Validate and print into console the correct output
void validate_card_number(long card_number, int length)
{
    if (calculate_checksum(card_number) % 10 == 0)
    {
        if (length == 16)
        {
            if (get_starting_numbers(card_number) == 51 ||
                get_starting_numbers(card_number) == 52 ||
                get_starting_numbers(card_number) == 53 ||
                get_starting_numbers(card_number) == 54 ||
                get_starting_numbers(card_number) == 55 ) printf("MASTERCARD\n");
            if (get_starting_numbers(card_number) >= 40 &&
                get_starting_numbers(card_number) <= 49) printf("VISA\n");
        }
        if (length == 15 && (get_starting_numbers(card_number) == 34 || get_starting_numbers(card_number) == 37 )) printf("AMEX\n");
        if (length == 13 && get_starting_numbers(card_number) == 4 ) printf("VISA\n");
    }
    else printf("INVALID\n");
}

File: Week-6-Python/sentimental-credit/credit.py
import re


def main():
    # Get card number.
    card_number = input("Number: ")

    # Validate card number.
    validate_card_number(card_number)


# Checksum.
def luhn(card_number):

    card_number = int(card_number)
    normal_sum = 0
    doubled_sum = 0
    counter = 0

    while card_number != 0:
        if counter % 2 == 0:
            normal_sum += card_number % 10
            card_number //= 10
            counter += 1
        else:
            temp_sum = (card_number % 10) * 2
            if temp_sum > 9:
                temp_sum -= 9
            doubled_sum += temp_sum
            card_number //= 10
            counter += 1

    return (normal_sum + doubled_sum) % 10 == 0


def validate_card_number(card_number):

    # Get card number length.
    card_number_length = len(card_number)

    if (not luhn(card_number)): print("INVALID")
    elif card_number[:2] in ["34", "37"] and card_number_length == 15: print("AMEX")
    elif card_number[:1] == "4" and (card_number_length == 13 or card_number_length == 16): print("VISA")
    elif card_number_length == 16 and re.match("5[12345]", card_number): print("MASTERCARD")
    else: print("INVALID")


if __name__ == "__main__":
    main()